packages/farrow-api-client/src/apiPipeline.ts

```diff
@@ -0,0 +1,137 @@
+import 'isomorphic-unfetch'
+import DataLoader from 'dataloader'
+import {
+  AsyncPipeline,
+  createAsyncPipeline,
+  getMiddleware,
+  MaybeAsync,
+  MiddlewareInput,
+  RunPipelineOptions,
+} from 'farrow-pipeline'
+import type {
+  ApiRequest,
+  Calling,
+  SingleCalling,
+  BatchCalling,
+  ApiResponse,
+  JsonType,
+  ApiErrorResponse,
+  ApiSuccessResponse,
+  ApiResponseSingle,
+} from 'farrow-api-server'
+
+export { ApiRequest, ApiResponse, JsonType, ApiErrorResponse, ApiSuccessResponse }
+
+export type ApiPipeline = AsyncPipeline<ApiRequest, ApiResponse> & {
+  match(pattern: string | RegExp, middleware: MiddlewareInput<ApiRequest, MaybeAsync<ApiResponse>>): void
+  invoke(url: string, calling: SingleCalling): Promise<JsonType | Error>
+  invoke(url: string, calling: BatchCalling): Promise<(JsonType | Error)[]>
+  invoke(url: string, calling: Calling): Promise<JsonType | Error | (JsonType | Error)[]>
+}
+
+export const createApiPipeline = (): ApiPipeline => {
+  let pipeline = createAsyncPipeline<ApiRequest, ApiResponse>()
+
+  function run(
+    request: ApiRequest,
+    options?: RunPipelineOptions<ApiRequest, MaybeAsync<ApiResponse>>,
+  ): MaybeAsync<ApiResponse> {
+    return pipeline.run(request, {
+      ...options,
+      onLast: fetcher,
+    })
+  }
+
+  let match: ApiPipeline['match'] = (pattern, middlewareInput) => {
+    let middleware = getMiddleware(middlewareInput)
+    pipeline.use((request, next) => {
+      if (pattern instanceof RegExp) {
+        if (pattern.test(request.url)) {
+          return middleware(request, next)
+        }
+        return next()
+      }
+
+      if (request.url.endsWith(pattern)) {
+        return middleware(request, next)
+      }
+
+      return next()
+    })
+  }
+
+  async function invoke(url: string, calling: SingleCalling): Promise<JsonType | Error>
+  async function invoke(url: string, calling: BatchCalling): Promise<(JsonType | Error)[]>
+  async function invoke(url: string, calling: Calling): Promise<JsonType | Error | (JsonType | Error)[]> {
+    let result = await run({ url, calling })
+
+    let handleResult = (apiResponse: ApiResponseSingle) => {
+      if (apiResponse.type === 'ApiErrorResponse') {
+        return new Error(apiResponse.error.message)
```

Comment:
Oh, I saw it here.

Why not `throw new Error` instead of `return new Error`

`async/await ` will take care the `throw error`, and turn it into `Promise.reject`.

User can `try-catch` the `errror` when `throw`.

Reply:
the result will be passed to DataLoader: 

```ts
let batchInvoke = (callings: Readonly<SingleCalling[]>) => {
  let body: BatchCalling = {
    __batch__: true,
    callings,
  }
  return apiPipeline.invoke(url, body)
}
let dataLoader = new DataLoader(batchInvoke)
```

and in dataloader should get error by result.

offical example:

```js
async function batchFunction(keys) {
  const results = await db.fetchAllKeys(keys)
  return keys.map(key => results[key] || new Error(`No result for ${key}`))
}

const loader = new DataLoader(batchFunction)
```

Reply:
got it